### ISSUE TITLE
Loss function

### ISSUE THREAD

**Issue by simon-p-j-r** [PlusLabNLP/AMPERE#2]
Dear author, thanks to the amazing work and open-sourced code. When I read your model_copyutils.py code,  I'm puzzled by some of your treatment of the loss function:
Why do you need to scatter_add() the last layer of cross_attention with lm_logits?
The implementation of the loss function in the code is skated over in the paper, could you please go into more detail?

**Comment by ihungalexhsu** [comment]
For the loss function:
1. https://github.com/PlusLabNLP/AMPERE/blob/main/src/model_copyutils.py#L198 I only select the last layer of cross attention in this line. Yet, as noted in the code, layers to aggregate can be changed here. (So basically, in the current code, line 199 only deals with "shape" but conceptually no particular usefulness.
2. https://github.com/PlusLabNLP/AMPERE/blob/main/src/model_copyutils.py#L200 takes average across different attention heads.
4. https://github.com/PlusLabNLP/AMPERE/blob/main/src/model_copyutils.py#L215 What this line does is: we add the `original_word_pro` (a distribution over all vocabulary) with [ input words (a sketchy distribution, only vocabs in input text have non-zero value) times cross-attention as weights ].

LMK if it's unclear.

**Comment by simon-p-j-r** [comment]
Thanks for your patience and replies.
I have two questions to consult with you.
1. In the paper, regularization loss (i.e. Eq. 3) does not introduce P^i_copy(j|·) but only W^i_gen. In introducing the above two parameters, the paper writes that P^i_copy(j|·) is related to cross_attention and W^i_gen is related to decoder_hidden_state. However, cross_attention is still introduced in the code implementation of regularization loss (i.e. Eq. 3), which I am very confused about.
![图片](https://github.com/PlusLabNLP/AMPERE/assets/67364150/228dc312-8e4b-4740-b292-8a64c22b9634)

2. I have implemented your DEGREE code and know that after training the completed model, the final data needs to be evaluated to get it. I also found the eval function [https://github.com/PlusLabNLP/AMPERE/blob/main/src/eval_EAE.py](url) in ampere's code, but there is no mention of the need to evaluate in ampere's Readme file, is it necessary for ampere to evaluate the trained model?

**Comment by ihungalexhsu** [comment]
Sorry for the confusion in the paper. To understand the total loss, 
You can see that 
![image](https://github.com/PlusLabNLP/AMPERE/assets/8895697/51fe7abf-f218-4ec4-8ba1-e92bcbc4100e) is from Eq.(2). and the loss regarding copy is written in Eq(2).

About evaluating AMPERE, you don't need to do so since AMPERE is for EAE.

**Comment by simon-p-j-r** [comment]
**Doesn't** ![图片](https://github.com/PlusLabNLP/AMPERE/assets/67364150/2aaa39b9-6f4a-4e3f-9987-a65f399afd3a) **come from** ![图片](https://github.com/PlusLabNLP/AMPERE/assets/67364150/97560d24-ba9b-4f7d-a391-c3cdc4dbe831) **but from** ![图片](https://github.com/PlusLabNLP/AMPERE/assets/67364150/80ac4122-c52a-4e0a-9eec-d8a7b89f1719)**?**

**Comment by ihungalexhsu** [comment]
Eq(1) is what DEGREE did. Eq(2) is applying copy mechanism. What we did in AMPERE is enhancing Eq(2) with regularization => Eq(3)

**Comment by simon-p-j-r** [comment]
I got it, thank you very much for your reply.
